### PR TITLE
Allow parsing any types in catch clauses in the parser.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This PR specifically makes it so that rules about ordering of catch clauses are not produced while parsing. Previously, we did this work at parse time. There's no reason for this complexity and specificity to be at the parser level. Syntactically we know what's going on, and we can allow this while parsing and check and report the error later. It also prevents incrementally reusing try-statements (and their containers) when this happens.

Tagging @dotnet/roslyn-compiler @gafter @VSadov

This part of a long term effort to remove unnecessary errors from the parser and move them to later stages of the compiler pipeline. Doing this makes the parser simpler, makes it faster to get a syntax tree, and makes our incremental implementation simpler (because the parser/incremental-parser needs to track less contextual states).

This is also important as syntax-diagnostics end up causing issues for higher level IDE features.

**Comment by CyrusNajmabadi** [comment]
Tagging @TyOverby and @jcouv as well.  Thanks!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -7722,18 +7722,17 @@ private CatchClauseSyntax ParseCatchClause(bool hasCatchAll)
...
-                    name = this.ParseIdentifierToken();
-                }
 
+                _termState |= TerminatorState.IsEndOfCatchClause;
+                var type = this.ParseType();
+                SyntaxToken name = this.IsTrueIdentifier()
+                    ? this.ParseIdentifierToken()
+                    : null;
                 _termState = saveTerm;
-                var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >var closeParen = this.EatToken(SyntaxKind.CloseParenToken); [](start = 16, length = 59)

Please restore the intermediate local.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -7722,18 +7722,17 @@ private CatchClauseSyntax ParseCatchClause(bool hasCatchAll)
...
-                if (this.IsTrueIdentifier())
-                {
-                    name = this.ParseIdentifierToken();
-                }
 
+                _termState |= TerminatorState.IsEndOfCatchClause;
+                var type = this.ParseType();
+                SyntaxToken name = this.IsTrueIdentifier()
+                    ? this.ParseIdentifierToken()
+                    : null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Please undo this style change.

**Comment by AlekseyTs** [comment]
Done with review pass.

**Comment by CyrusNajmabadi** [comment]
Code updated.  @dotnet/roslyn-compiler  plz review.

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs Can you take another look?  Thanks!

**Comment by AlekseyTs** [comment]
LGTM

**Pull request merged**